### PR TITLE
io_uring: add client socket

### PR DESCRIPTION
**Pull request by zhxie**
Signed-off-by: He Jie Xu <hejie.xu@intel.com>
Signed-off-by: Xie Zhihao <zhihao.xie@intel.com>

Commit Message: io_uring: add client socket
Additional Description: 
Risk Level: Low
Testing: Unit and integration tests, #28824
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A
Part of Issue #28395 
Continue of #28706 #30773

### CODE REVIEW

File: envoy/common/io/io_uring.h [hunk 1]
```diff
@@ -254,10 +254,16 @@ class IoUringSocket {
 
   /**
    * Enable close event. This is used for the case the socket is listening on the file close event.
-   * Then a remote close is found by a read request will delievered as file close event.
+   * Then a remote close is found by a read request will delivered as file close event.
```
Comment (soulxu) [inline envoy/common/io/io_uring.h:257]: emm... seems like it should be `will be delivered`?

File: test/common/io/io_uring_worker_impl_integration_test.cc [hunk 2]
```diff
@@ -877,6 +884,74 @@ TEST_F(IoUringWorkerIntegrationTest, ServerSocketUpdateFileEventCb) {
...
+        EXPECT_NE(absl::nullopt, socket->getWriteParam());
+        result = socket->getWriteParam()->result_;
+      },
+      false);
+  EXPECT_EQ(io_uring_worker_->getSockets().size(), 1);
+
+  // Waiting for the client socket connect.
+  struct sockaddr_in listen_addr = getListenSocketAddress();
+  auto addr = std::make_shared<Network::Address::Ipv4Instance>(&listen_addr);
+  io_uring_worker_->submitConnectRequest(*socket, addr);
```
Comment (soulxu) [inline test/common/io/io_uring_worker_impl_integration_test.cc:906]: we should use the `socket->connect`

File: test/common/io/io_uring_worker_impl_integration_test.cc [hunk 3]
```diff
@@ -877,6 +884,74 @@ TEST_F(IoUringWorkerIntegrationTest, ServerSocketUpdateFileEventCb) {
...
+        ASSERT(events == Event::FileReadyType::Write);
+        EXPECT_NE(absl::nullopt, socket->getWriteParam());
+        result = socket->getWriteParam()->result_;
+      },
+      false);
+  EXPECT_EQ(io_uring_worker_->getSockets().size(), 1);
+
+  // Waiting for the client socket connect.
+  auto addr = std::make_shared<Network::Address::Ipv4Instance>(0);
+  io_uring_worker_->submitConnectRequest(*socket, addr);
```
Comment (soulxu) [inline test/common/io/io_uring_worker_impl_integration_test.cc:941]: ditto, we should use `socket->connect()`

File: source/common/io/io_uring_worker_impl.cc [hunk 4]
```diff
@@ -611,5 +638,41 @@ void IoUringServerSocket::submitWriteOrShutdownRequest() {
...
+  if (status_ == Closed) {
+    if (write_or_shutdown_req_ == nullptr && read_cancel_req_ == nullptr &&
+        write_or_shutdown_cancel_req_ == nullptr) {
+      closeInternal();
+    }
+    return;
+  }
+
+  if (result == 0) {
+    enableRead();
```
Comment (soulxu) [inline source/common/io/io_uring_worker_impl.cc:670]: maybe we can add unittest for the case the connection failed and the read won't be enabled.

File: source/common/io/io_uring_worker_impl.cc [hunk 5]
```diff
@@ -611,5 +638,41 @@ void IoUringServerSocket::submitWriteOrShutdownRequest() {
   }
 }
 
+IoUringClientSocket::IoUringClientSocket(os_fd_t fd, IoUringWorkerImpl& parent,
+                                         Event::FileReadyCb cb, uint32_t write_timeout_ms,
+                                         bool enable_close_event)
+    : IoUringServerSocket(fd, parent, cb, write_timeout_ms, enable_close_event) {}
+
+void IoUringClientSocket::connect(const Network::Address::InstanceConstSharedPtr& address) {
+  // Reuse read request since there is no read on connecting and connect is cancellable.
```
Comment (soulxu) [inline source/common/io/io_uring_worker_impl.cc:647]: I wasn't sure reuse the `read_req_` is good choice. That may tricky us in the future.

**Review by soulxu — CHANGES_REQUESTED** [review]
thanks! few comments

File: source/common/io/io_uring_worker_impl.cc (hunk 4 above)
Comment (zhxie) [reply source/common/io/io_uring_worker_impl.cc:670]: Added, thanks.

File: source/common/io/io_uring_worker_impl.cc (hunk 5 above)
Comment (zhxie) [reply source/common/io/io_uring_worker_impl.cc:647]: read and connect are consistent in the current logical handling (in close, cancel). Given the simplicity of the code, we are going to maintain the status after discussing it offline.

Comment (soulxu) [reply source/common/io/io_uring_worker_impl.cc:647]: yes, let us keep it.

**Review by soulxu — APPROVED** [review]
LGTM, thanks! waiting @yanavlasov for senior maintainer review

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**